src/OpenTelemetry/Logs/LogProcessor.cs

```diff
@@ -0,0 +1,185 @@
+// <copyright file="LogProcessor.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#if NETSTANDARD2_0
+using System;
+using System.Threading;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Logs
+{
+    /// <summary>
+    /// Log processor base class.
+    /// </summary>
+    public abstract class LogProcessor : IDisposable
+    {
+        private int shutdownCount;
+
+        /// <summary>
+        /// Called synchronously when a <see cref="LogRecord"/> is recorded.
+        /// </summary>
+        /// <param name="record">
+        /// The log record.
+        /// </param>
+        /// <remarks>
+        /// This function is called synchronously on the thread which generated
+        /// the log record. This function should be thread-safe, and should not
+        /// block indefinitely or throw exceptions.
+        /// </remarks>
+        public virtual void OnLog(in LogRecord record)
+        {
+        }
+
+        /// <summary>
+        /// Flushes the <see cref="LogProcessor"/>, blocks the current
+        /// thread until flush completed, shutdown signaled or timed out.
+        /// </summary>
+        /// <param name="timeoutMilliseconds">
+        /// The number of milliseconds to wait, or <c>Timeout.Infinite</c> to
+        /// wait indefinitely.
+        /// </param>
+        /// <returns>
+        /// Returns <c>true</c> when flush succeeded; otherwise, <c>false</c>.
+        /// </returns>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the <c>timeoutMilliseconds</c> is smaller than -1.
+        /// </exception>
+        /// <remarks>
+        /// This function guarantees thread-safety.
+        /// </remarks>
+        public bool ForceFlush(int timeoutMilliseconds = Timeout.Infinite)
+        {
+            if (timeoutMilliseconds < 0 && timeoutMilliseconds != Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeoutMilliseconds));
```

Comment:
Include the actual value which is out of range into the exception message?

Reply:
I like the idea, while it is not the [practice in this repo](https://github.com/open-telemetry/opentelemetry-dotnet/search?p=1&q=ArgumentOutOfRangeException).

@cijothomas @CodeBlanch @eddynaka do you think this is something we would want (e.g. "xyz is out of range, the expect range is foo, the actual value is bar")?

Reply:
I think you can specify a message along with the name a la:
`throw new ArgumentOutOfRangeException(nameof(timeoutMilliseconds), "Timeout should be a positive value or zero.");`
There might even be an overload that takes name, value, & message.

It's a bit tedious to do that level of verbosity but I'm sure users would appreciate the detail when they do encounter an error.

Reply:
I would add the message to help developers as well. At first, just add in English and, later, add based on culture.

Reply:
Created issue #1316 to track this across the repo.